diff --git a/makefile b/makefile
index 6aaf70d..7567358 100644
--- a/makefile
+++ b/makefile
@@ -25,6 +25,15 @@ CFLAGS += \
 	-DCOMPAT_PROJECT="\"Compat-rdma\"" \
 	-DCOMPAT_VERSION="\"$(shell cat compat_version)\"" \
 
+ifneq ($(CONFIG_INFINIBAND_SCIF)$(CONFIG_IBP_SERVER),)
+SCIF_HEADER3X:=/usr/src/kernels/$(KVERSION)/include/modules/
+SCIF_HEADER4X:=/usr/src/mpss-modules/
+CFLAGS += -I$(SCIF_HEADER3X)
+CFLAGS += -I$(SCIF_HEADER4X)
+CFLAGS += \
+       $(shell if grep -R -q scif_get_node_ids $(SCIF_HEADER3X)/* $(SCIF_HEADER4X)/* 2>/dev/null; then echo -n -DHAVE_UPSTREAM_SCIF; fi)
+endif
+
 DEPMOD  = /sbin/depmod
 INSTALL_MOD_DIR ?= $(shell test -f /etc/redhat-release && echo extra/ofa_kernel || echo updates)
 
